File: 2018/day/19/Main.hs
{-# LANGUAGE ViewPatterns, RecordWildCards, LambdaCase #-}
import Control.Monad
import Data.Array
import Data.Bits
import Data.Char
import Data.Either
import Data.List
import Data.Map (Map)
import qualified Data.Map as Map
import Data.Maybe
import Text.ParserCombinators.ReadP
import Text.Printf

import Debug.Trace

data Instruction = I String Int Int Int deriving (Eq)

instance Show Instruction where
    show (I opname a b c) = printf "%s %d %d %d" opname a b c

data Device = Dev
    { boundIp :: Int
    , ip :: Int
    , reg :: Array Int Int
    , code :: Array Int Instruction
    } deriving (Show)

parse = do
    boundIp <- string "#ip " *> parseNum <* char '\n'
    xs <- sepBy1 parseInstruction (char '\n')
    optional (char '\n')
    eof
    return $ Dev boundIp 0 (listArray (0, 5) (repeat 0)) (listArray (0, length xs - 1) xs)

parseNum = read <$> munch1 isDigit
   
parseInstruction = do
    opname <- count 4 get
    [a, b, c] <- count 3 (char ' ' *> parseNum)
    return $ I opname a b c
              
halted Dev{..} = not (inRange (bounds code) ip)

eval Dev{..} =
    case opname of
        "addr" -> step $ \reg a b c -> reg // [(c, reg ! a + reg ! b)]
        "addi" -> step $ \reg a b c -> reg // [(c, reg ! a + b)]
        "mulr" -> step $ \reg a b c -> reg // [(c, reg ! a * reg ! b)]
        "muli" -> step $ \reg a b c -> reg // [(c, reg ! a * b)]
        "banr" -> step $ \reg a b c -> reg // [(c, reg ! a .&. reg ! b)]
        "bani" -> step $ \reg a b c -> reg // [(c, reg ! a .&. b)]
        "borr" -> step $ \reg a b c -> reg // [(c, reg ! a .|. reg ! b)]
        "bori" -> step $ \reg a b c -> reg // [(c, reg ! a .|. b)]
        "setr" -> step $ \reg a b c -> reg // [(c, reg ! a)]
        "seti" -> step $ \reg a b c -> reg // [(c, a)]
        "gtir" -> step $ \reg a b c -> reg // [(c, fromEnum $ a > reg ! b)]
        "gtri" -> step $ \reg a b c -> reg // [(c, fromEnum $ reg ! a > b)]
        "gtrr" -> step $ \reg a b c -> reg // [(c, fromEnum $ reg ! a > reg ! b)]
        "eqir" -> step $ \reg a b c -> reg // [(c, fromEnum $ a == reg ! b)]
        "eqri" -> step $ \reg a b c -> reg // [(c, fromEnum $ reg ! a == b)]
        "eqrr" -> step $ \reg a b c -> reg // [(c, fromEnum $ reg ! a == reg ! b)]
  where
    I opname a b c = code ! ip
    step f = let reg' = f (reg // [(boundIp, ip)]) a b c
             in Dev boundIp ((reg' ! boundIp) + 1) reg' code

codegen :: Device -> String
codegen Dev{..} = unlines $
    [
      "#include <stdio.h>"
    , ""
    , "int main(int argc, char **argv) {"
    , "  int r0 = 0, r1 = 0, r2 = 0, r3 = 0, r4 = 0, r5 = 0;  // Registers"
    , "  int *ip = &r1;  // Bind instruction pointer to register 1"
    , ""
    , "  while (*ip < " ++ show n ++ ") {"
    , "    switch (*ip) {"
    ]
    ++ map ("      " ++ ) (map genCase (assocs code))
    ++
    [
      "    }"
    , ""
    , "    *ip += 1;"
    , "  }"
    , ""
    , "  printf(\"%d\\n\", r0);"
    , "}"
    ]
  where
    n = length code
    
    f = \case
        I "addr" a b c -> printf "r%d = r%d + r%d;" c a b
        I "addi" a b c -> printf "r%d = r%d + %d;" c a b
        I "mulr" a b c -> printf "r%d = r%d * r%d;" c a b
        I "muli" a b c -> printf "r%d = r%d * %d;" c a b
        I "banr" a b c -> printf "r%d = r%d & r%d;" c a b
        I "bani" a b c -> printf "r%d = r%d & %d;" c a b
        I "borr" a b c -> printf "r%d = r%d | r%d;" c a b
        I "bori" a b c -> printf "r%d = r%d | %d;" c a b
        I "setr" a b c -> printf "r%d = r%d;" c a
        I "seti" a b c -> printf "r%d = %d;" c a
        I "gtir" a b c -> printf "r%d = %d > r%d;" c a b
        I "gtri" a b c -> printf "r%d = r%d > %d;" c a b
        I "gtrr" a b c -> printf "r%d = r%d > r%d;" c a b
        I "eqir" a b c -> printf "r%d = %d == r%d;" c a b
        I "eqri" a b c -> printf "r%d = r%d == %d;" c a b
        I "eqrr" a b c -> printf "r%d = r%d == r%d;" c a b

    genCase (i, x@(I op a b c)) =
      "case " ++ show i ++ ": " ++ f x ++ " break;"

main = do
    input <- readFile "input.txt"
    let [(dev0, "")] = readP_to_S parse input

    -- Print C version of the program
    -- putStrLn (codegen dev0)

    -- Part 1
    print $ (! 0) . reg $ until halted eval dev0

    -- Part 2
    -- let dev1 = dev0 { reg = reg dev0 // [(0, 1)] }
    -- print $ (! 0) . reg $ until halted eval dev0
  
    -- Print C version of the program
    -- putStrLn (codegen dev0)
    let r5 = 10551296
    print $ r5 + sum [d | d <- [1..r5 `div` 2], r5 `mod` d == 0]


    -- mapM_ (uncurry debug) $ sample 10551296 $ zip [0..] $ iterate eval dev1

    -- plotRun dev1

File: 2018/day/19/main.c
#include <stdio.h>

int main(int argc, char **argv) {
    int n = 0;
    int r0 = 0, r1 = 0, r2 = 0, r3 = 0, r4 = 0, r5 = 0;
    r0 = 1;
    int *ip = &r1;
    
    while (*ip < 36) {
        printf("n=%d,ip=%d,r0=%d,r1=%d,r2=%d,r3=%d,r4=%d,r5=%d\n", n, *ip, r0, r1, r2, r3, r4, r5);

        switch (*ip) {
            case 0: r1 = r1 + 16; break;
            case 1: r2 = 1; break;
            case 2: r3 = 1; break;
            case 3: r4 = r2 * r3; break;
            case 4: r4 = r4 == r5; break;
            case 5: r1 = r4 + r1; break;
            case 6: r1 = r1 + 1; break;
            case 7: r0 = r2 + r0; break;
            case 8: r3 = r3 + 1; break;
            case 9: r4 = r3 > r5; break;
            case 10: r1 = r1 + r4; break;
            case 11: r1 = 2; break;
            case 12: r2 = r2 + 1; break;
            case 13: r4 = r2 > r5; break;
            case 14: r1 = r4 + r1; break;
            case 15: r1 = 1; break;
            case 16: r1 = r1 * r1; break;
            // init
            case 17: r5 = r5 + 2; break; // r5 = 2
            case 18: r5 = r5 * r5; break; // r5 = 4
            case 19: r5 = r1 * r5; break; // r5 = 76
            case 20: r5 = r5 * 11; break; // r5 = 836
            case 21: r4 = r4 + 2; break; // r4 = 2
            case 22: r4 = r4 * r1; break; // r4 = 44
            case 23: r4 = r4 + 16; break; // r4 = 60
            case 24: r5 = r5 + r4; break; // r5 = 896
            case 25: r1 = r1 + r0; break;
            case 26: r1 = 0; break; // goto 1 if r0 = 0
            case 27: r4 = r1; break; // here r1 = 1, r4=27
            case 28: r4 = r4 * r1; break; // r4 *= 28
            case 29: r4 = r1 + r4; break; // r4 += 29
            case 30: r4 = r1 * r4; break; // r4 *= 30
            case 31: r4 = r4 * 14; break; // r4 *= 14
            case 32: r4 = r4 * r1; break; // r4 *= 32
            case 33: r5 = r5 + r4; break; // r5 = ((27*28)+29)*30*14*32+896
            case 34: r0 = 0; break;
            case 35: r1 = 0; break;
        }

        *ip += 1;
        n++;
        printf("n=%d,ip=%d,r0=%d,r1=%d,r2=%d,r3=%d,r4=%d,r5=%d\n", n, *ip, r0, r1, r2, r3, r4, r5);
        getchar();
    }

    printf("r0 = %d\n", r0);
}